activerecord/lib/active_record/relation.rb

```diff
@@ -170,6 +159,42 @@ def find_or_create_by!(attributes, &block)
       find_by(attributes) || create!(attributes, &block)
     end
 
+    # Attempts to create a record with the given attributes in a table that has a unique constraint
+    # on one or several of its columns. If a row already exists with one or several of these
+    # unique constraints, the exception such an insertion would normally raise is caught, 
+    # and the existing record with those attributes is sought found using #find_by.
+    #
+    # This is similar to #find_or_create_by, but avoids the problem of stale reads, as that methods needs
+    # to first query the table, then attempt to insert a row if none is found. That leaves a timing gap
+    # between the SELECT and the INSERT statements that can cause problems in high throughput applications.
+    #
+    # There are several drawbacks to #create_or_find_by, though:
+    #
+    # * The underlying table must have the relevant columns defined with unique constraints.
+    # * A unique constraint violation may be triggered by only one, or at least less than all, 
+    #   of the given attributes. This means that the subsequent #find_by may fail to find a
+    #   matching record, which will then raise an `ActiveRecord::NotFound` exception,
+    #   rather than a record will the given attributes.
+    # * It relies on exception handling to handle control flow, which may be marginally slower. And
+    #
+    # This method will always returns a record if all given attributes are covered by unique constraints,
+    # but if creation was attempted and failed due to validation errors it won't be persisted, you get what
+    # #create returns in such situation.
+    def create_or_find_by(attributes, &block)
+      create(attributes, &block)
+    rescue ActiveRecord::RecordNotUnique
```

Comment:
This needs `transaction(requires_new: true) do` around the `create` to work in an ongoing surrounding transaction (on at least PostgreSQL)

Reply:
Why is this not necessary for find_or_create_by?

Reply:
Because it doesn't cause an SQL error and then attempt to recover. PostgreSQL remembers when an error has occurred inside a transaction, and disallows all further operations until that transaction has been rolled back.

Reply:
If we are willing to require PG 9.5 or later for this (when using the PG adapter), we could just stick `ON CONFLICT DO NOTHING` on the end instead. This would let us wrap the whole thing in a transaction instead of just the create, which eliminates any possibility of race conditions.